features/distributed/kv-store/postgres-shared/src/main/java/org/opennms/features/distributed/kvstore/pgshared/AbstractPostgresKeyValueStore.java

```diff
@@ -0,0 +1,250 @@
+ * @param <S> the SQL type this store persists if the value type must be wrapped with an SQL type
+ */
+public abstract class AbstractPostgresKeyValueStore<T, S> extends AbstractAsyncKeyValueStore<T> {
+    private static final String VALUE_COLUMN = "value";
+    private static final String KEY_COLUMN = "key";
+    private static final String CONTEXT_COLUMN = "context";
+    private static final String LAST_UPDATED_COLUMN = "last_updated";
+    private static final String EXPIRES_AT_COLUMN = "expires_at";
+
+    private final PreparedStatement selectStatement;
+    private final PreparedStatement upsertStatement;
+    private final PreparedStatement lastUpdatedStatement;
+
+    public AbstractPostgresKeyValueStore(DataSource dataSource) {
+        Objects.requireNonNull(dataSource);
+
+        try {
+            Connection connection = dataSource.getConnection();
+
+            // Prepare the statements now that we have the table name
+            selectStatement = connection.prepareStatement(String.format("SELECT %s, %s FROM %s WHERE %s = ? AND %s = ?",
+                    VALUE_COLUMN, EXPIRES_AT_COLUMN, getTableName(), KEY_COLUMN, CONTEXT_COLUMN));
+            upsertStatement = connection.prepareStatement(String.format(
+                    "INSERT INTO %s (%s, %s, %s, %s, %s) VALUES (?, ?, ?, ?, " + getValueStatementPlaceholder() + ") " +
+                            "ON CONFLICT ON CONSTRAINT " +
+                            getPkConstraintName() +
+                            " DO " +
+                            "UPDATE SET %s = ?, %s = ?, %s = " + getValueStatementPlaceholder(),
+                    getTableName(), KEY_COLUMN, CONTEXT_COLUMN, LAST_UPDATED_COLUMN, EXPIRES_AT_COLUMN, VALUE_COLUMN,
+                    LAST_UPDATED_COLUMN, EXPIRES_AT_COLUMN, VALUE_COLUMN
+            ));
+            lastUpdatedStatement = connection.prepareStatement(String.format("SELECT %s, %s FROM %s WHERE %s = ? AND " +
+                            "%s =" +
+                            " ?",
+                    LAST_UPDATED_COLUMN, EXPIRES_AT_COLUMN, getTableName(), KEY_COLUMN, CONTEXT_COLUMN));
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static boolean expired(ResultSet resultSet) throws SQLException {
+        long now = System.currentTimeMillis();
+        Timestamp expiresAt = resultSet.getTimestamp(EXPIRES_AT_COLUMN);
+
+        if (expiresAt != null && expiresAt.getTime() < now) {
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public long put(String key, T value, String context, Integer ttlInSeconds) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+
+        long now = System.currentTimeMillis();
+
+        try {
+            // The below sets the prepared values for both the INSERT and UPDATE cases hence some values being repeated
+            upsertStatement.setString(1, key);
```

Comment:
Are these prepared statements and the connection thread safe?

Reply:
No, good catch. I sent them back to thread safety training, think they should be good now.